includes/handler.inc

```diff
@@ -430,46 +473,184 @@ function islandora_oai_query_solr($start = 0, $field = 'PID', $query = NULL, $se
+          foreach ($sets as $value) {
+            $query_results = islandora_oai_query_solr_for_child_sets($value);
+            foreach ($query_results as $result) {
+              if (!in_array($result, $descendants)) {
+                $descendants[] = $result;
+              }
+            }
+            $descendants[] = $value;
+          }
+          $query_processor->solrParams['fq'][] = islandora_oai_build_descendants_fq_statement($descendants);
+        }
+        break;
     }
-    else {
-      $user_name = $user->name;
+  }
+  $query_processor->solrStart = $start;
+  $query_processor->solrLimit = variable_get('islandora_oai_max_size', '20');
+  try {
+    $query_processor->executeQuery(FALSE);
+    $solr_results = $query_processor->islandoraSolrResult['response'];
+  }
+  catch (Exception $e) {
+    drupal_set_message(t('error searching @message', array('@message' => $e->getMessage())), 'error');
+  }
+  return $solr_results;
+}
+
+/**
+ * Helper function to build an 'fq' statement from a descentants array.
+ *
+ * @param array $descendants
+ *   The array of descendant collection PIDs.
+ *
+ * @return string
+ *   The 'fq' statement.
+ */
+function islandora_oai_build_descendants_fq_statement(array $descendants) {
+  $set_fq = array();
+  // We are using OR here to account for fields in Solr that may index
+  // just the PID or the entire URI. In the future if performance becomes
+  // an issue with huge Solr queries we should re-visit this.
+  foreach ($descendants as $walk) {
+    foreach (islandora_oai_get_membership_array() as $collection_field) {
+      $set_fq[] = $collection_field . ':("' . $walk . '" OR "info:fedora/' . $walk . '")';
     }
+  }
+  return '(' . implode(' OR ', $set_fq) . ')';
+}
 
-    // Get all descendant children (collection and objs) and add to the sets
-    // array if not already existing.
-    // Cannot constrain on collections only as name conventions may change
-    // across implementations.
-    foreach ($sets as $key => $value) {
-      // Replace the first occurrence of _ with a : so we are back to a
-      // valid pid.
-      $value = preg_replace('/_/', ':', $value, 1);
-      $value = trim($value);
-      $sets[$key] = $value;
+/**
+ * Helper function to query recursively for child sets from Solr.
+ *
+ * @param string $parent
+ *   The set to get children in.
+ * @param array $sets
+ *   Used internally to maintain the list of sets found as it recurses.
+ *
+ * @return array
+ *   A list of set PIDs.
+ */
+function islandora_oai_query_solr_for_child_sets($parent, &$sets = array()) {
+  global $user;
+  $user_name = $user->uid === 0 ? 'anonymous' : $user->name;
+
+  module_load_include('inc', 'islandora_solr', 'includes/utilities');
+  $qp = new IslandoraSolrQueryProcessor();
+  $qp->buildQuery('*:*');
+  $qp->solrParams['fl'] = 'PID';
+  // Make sure we get everything but still filter on XACML permissions.
+  $qp->solrParams['fq'] = array();
+  if (module_exists('islandora_xacml_api')) {
```

Comment:
Apparently, the `module_exists()` call isn't necessary, since [`module_invoke()`](https://api.drupal.org/api/drupal/includes%21module.inc/function/module_invoke/7.x) calls [`module_hook()`](https://api.drupal.org/api/drupal/includes%21module.inc/function/module_hook/7.x), which will prevent the call to the non-existent function... Shrug?